File: src/llm_cal/output/labels.py
"""6-level label discipline — the soul of the tool.

Every number in the output must be wrapped in `AnnotatedValue` so users always know
where a value came from. Using `StrEnum` (not bare strings) means typos are caught by
mypy/ruff, not by users.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum
from typing import Generic, TypeVar


class Label(StrEnum):
    VERIFIED = "verified"
    INFERRED = "inferred"
    ESTIMATED = "estimated"
    CITED = "cited"
    UNVERIFIED = "unverified"
    UNKNOWN = "unknown"
    # Experimental opt-in 7th level. Populated only when --llm-review is used.
    # Never overrides the first 6 — it's an external second opinion, not truth.
    LLM_OPINION = "llm-opinion"


T = TypeVar("T")


@dataclass(frozen=True)
class AnnotatedValue(Generic[T]):
    """A value paired with provenance metadata.

    Examples:
        AnnotatedValue(160_300_000_000, Label.VERIFIED, source="HF model_info.siblings")
        AnnotatedValue(4.52, Label.INFERRED, source="160.3 GB / 284B params")
        AnnotatedValue(2_600_000_000, Label.ESTIMATED,
                       source="compress_ratios=[0,0,4,128,...] at 128K ctx")
    """

    value: T
    label: Label
    source: str | None = None

    def render_tag(self) -> str:
        return f"[{self.label.value}]"

File: crates/llm-infer-cal-core/src/output/labels.rs
use std::fmt;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Label {
    Verified,
    Inferred,
    Estimated,
    Cited,
    Unverified,
    Unknown,
    LlmOpinion,
}

impl Label {
    pub const fn all() -> [Label; 7] {
        [
            Label::Verified,
            Label::Inferred,
            Label::Estimated,
            Label::Cited,
            Label::Unverified,
            Label::Unknown,
            Label::LlmOpinion,
        ]
    }

    pub const fn as_str(self) -> &'static str {
        match self {
            Label::Verified => "verified",
            Label::Inferred => "inferred",
            Label::Estimated => "estimated",
            Label::Cited => "cited",
            Label::Unverified => "unverified",
            Label::Unknown => "unknown",
            Label::LlmOpinion => "llm-opinion",
        }
    }
}

impl fmt::Display for Label {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct AnnotatedValue<T> {
    pub value: T,
    pub label: Label,
    pub source: Option<String>,
}

impl<T> AnnotatedValue<T> {
    pub fn new(value: T, label: Label, source: Option<&str>) -> Self {
        Self {
            value,
            label,
            source: source.map(str::to_string),
        }
    }

    pub fn render_tag(&self) -> String {
        format!("[{}]", self.label.as_str())
    }
}
